yktrace/src/hwt/mapper.rs

```diff
@@ -183,7 +183,12 @@ impl HWTMapper {
     fn map_block(&mut self, block: &hwtracer::Block) -> Vec<IRBlock> {
         let block_vaddr = block.first_instr();
         let (obj_name, block_off) = code_vaddr_to_off(block_vaddr as usize).unwrap();
-        let block_len = block.last_instr() - block_vaddr;
+
+        // The `+ 1` here is because hwtracer gives us the address of the last instruction in the
```

Comment:
What happens if the last instruction is more than 1 byte long?

Reply:
It doesn't matter. As long as at least one byte of an instruction is captured, the correct thing will happen.

This ickyness is all born out of a quirk in libipt. It can only report the address of the last instruction in a PT block and not the last byte of the block.

Reply:
I think this is crucial "As long as at least one byte of an instruction is captured, the correct thing will happen." I think the comment at the moment is a bit wooly, but that sentence captures something important. One thing I'm not sure about is what "the correct thing" is, but maybe that's me being an idiot. Can we rephrase the comment capture the crucial bit?

Reply:
Yes we can. Let me try to explain it here first, as I think what I said above isn't quite correct. Hopefully then we can boil the comment down to something concise and accurate.

It's to do with the way we store the [block map](https://llvm.org/docs/Extensions.html#sht-llvm-bb-addr-map-section-basic-block-address-map). We use an interval tree which stores a collection of machine basic block ranges. When we get a PT block, we want to know which machine block(s) that block maps to. An interval tree is perfect for this, as you can query it with a range and any and all overlapping ranges are returned.

Let's run some examples. Suppose we assign each instruction in a block with a letter, and that each instance of the letter represents a byte of that instruction. Let's use a pipe to delimit machine blocks.

Suppose that the 4 machine blocks of a function are:
```
abbcc|ddee|f|gghh
0     5    9 10
```

[the numbers underneath are the start address of the machine block]

If libipt gives us a `(block-addr, last-instr-addr)` pair. So if we got (0, 7) from libipt, then we expect to get the first two blocks. The code in-tree would get this right even though it would query that range with an exclusive upper bound.

Where the in-tree code breaks down is when the block is one-byte long. So if we got (9, 9), then we'd query the range `9..9` which is an empty range. By definition, querying an interval tree with an empty range always yields the empty set, which means we erroneously fail to map the block.

What the `+1` does is it ensures that the upper bound is exclusive. Then the 1-byte block is correctly mapped. We could have said "if the block length is 1, then do a query with an exclusive upper bound", but there's no need. Since every instruction is at least one byte long, adding one will not break the cases where the length of the block is longer than a single byte. Therefore adding one is always correct (you can never spill into the next block and report one too many blocks).

Therefore:
> As long as at least one byte of an instruction is captured, the correct thing will happen.

Is wrong. It should read:
> As long as at least one byte of the machine block is captured, the correct thing will happen.

In other words, in order for the correct blocks to appear in the trace, we must query the interval tree with a range that overlaps with at least one byte of each corresponding machine basic block.

Does that make sense? @ptersilie how about you?

Reply:
OK, I think I understand. How about something like "The HW mapper gives us the start address of the last instruction in a block, which works OK except for blocks with a single 1 byte instruction for which we generate an interval `(x,x)` i.e. an empty interval. Fortunately we don't need to calculate the precise end address of the last instruction in a block (on variable width instruction sets that's impractical!): instead we just need to ensure we're at least 1 byte into that last instruction (hence the `+1` below). Since every instruction is at least 1 byte long, this is always safe, and ensures that we always generate a valid interval."

Reply:
Almost.

> we just need to ensure we're at least 1 byte into that last instruction

We just need to make sure we are one byte into the last machine block.

Do you agree?

Reply:
> We just need to make sure we are one byte into the last machine block.

I don't think "last machine block" makes sense (last machine block relative to ... ?). I *think* "last instruction" is correct (at least if I've understood what the HW mapper and interval tree do).

Reply:
I think you are right actually.

Reply:
OK. Shall we use my suggestion above (but with perhaps "s/single 1 byte instruction/that consist of 1 instruction and where that instruction is 1 byte long/")?

Reply:
Reading this again, I think this could be more objective:

How about:
> We need to check which LLVM machine block(s) the hardware block coincides with (whose intervals overlap). Ideally we'd query the entire range of the hardware block, but libipt can only tell us the range up until the beginning of the last CPU instruction in the block. This is actually fine, as for a machine block to match, at least one-byte must overlap with the queried range. Therefore we can safely query the range libipt gives us (just ensuring that the upper bound is inclusive so that 1-byte blocks are captured).

Reply:
I again fear that I am once again slipping into arrogance, but this seems both less precise and less easy to read than my suggestion?

Reply:
Let me try to win you over to my way of thinking:

Our query interval must overlap each of the corresponding machine blocks by at least one byte. This is the weakest sufficient condition for correctly mapping blocks.

Suppose that we have a query that should map to the first 3 blocks of the following blockmap layout:

```
blockmap: |-----|---|-|----|
expected: |-----|---|-|       This is the correct mapping result.
query1  : |-----------|       OK and exact.
query2  :     |-------|       OK, overlaps all expected blocks too.
query3  :       |-----|       Bad, no byte in the query overlaps with the first block.  
query4  : |---------|         Bad, no byte in the query overlaps with the third block.
```

I find this way of thinking much more intuitive.

Note also that:

> which works OK except for blocks with a single 1 byte instruction for which we generate an interval (x,x) i.e. an empty interval

Isn't quite right. The scenario we are guarding against is when the last block in a query is a 1-byte block. Query 4 serves as a counter-example. The query is not empty, but the 3rd block is missed because the interval bound lacked an exclusive upper bound.

Makes sense?

Reply:
All of this to say, we just need to ensure the upper bound is an inclusive bound :P

Reply:
How about merging the two into something like:

"The HW mapper gives us the start address of the last instruction in a block, which gives us an exclusive upper bound when the interval tree expects an inclusive upper bound. This is a problem when the last block in our sequence has a single 1 byte instruction: the exclusive upper bound will not include any of that block's bytes. Fortunately, we don't need to calculate the precise end address of the last instruction in a block (on variable width instruction sets that's impractical!): instead we just need to ensure we're at least 1 byte into that last instruction (hence the +1 below). Since every instruction (and thus every block) is at least 1 byte long, this is always safe, and ensures that we generate an inclusive upper bound."

Reply:
Jolly good. Fixed in 708e8e4d4609.